src/tactic/solve_by_elim.lean

```diff
@@ -26,35 +26,60 @@ namespace tactic
+objects now.
+
+As an example, we have `def rfl : ∀ {α : Sort u} {a : α}, a = a`, which on elaboration will become
+`@rfl ?m_1 ?m_2`.
+
+Because `solve_by_elim` works by repeated application of lemmas against subgoals,
+the first time such a lemma is successfully applied,
+those metavariables will be unified, and thereafter have fixed values.
+This would make it impossible to apply the lemma
+a second time with different values of the metavariables.
+
+See https://github.com/leanprover-community/mathlib/issues/2269
+
+As an optimisation, after we build the list of `tactic expr`s, we actually run them, and replace any
+that do not in fact produce metavariables with a simple `return` tactic.
 -/
 meta def mk_assumption_set (no_dflt : bool) (hs : list simp_arg_type) (attr : list name) :
-  tactic (list expr) :=
-do (hs, gex, hex, all_hyps) ← decode_simp_arg_list hs,
-   hs ← hs.mmap i_to_expr_for_apply,
-   l ← attr.mmap $ λ a, attribute.get_instances a,
-   let l := l.join,
-   m ← list.mmap mk_const l,
-   let hs := (hs ++ m).filter $ λ h, expr.const_name h ∉ gex,
-   hs ← if no_dflt then
-          return hs
-        else
-          do { rfl_const ← mk_const `rfl,
-               trivial_const ← mk_const `trivial,
-               congr_fun ← mk_const `congr_fun,
-               congr_arg ← mk_const `congr_arg,
-               return (rfl_const :: trivial_const :: congr_fun :: congr_arg :: hs) },
-   if ¬ no_dflt ∨ all_hyps then do
+  tactic (list (tactic expr)) :=
+ -- We lock the tactic state so that any spurious goals generated during
+ -- elaboration of pre-expressions are discarded
+ lock_tactic_state $
+do
+  -- `hs` are expressions specified explicitly,
+  -- `hex` are exceptions (specified via `solve_by_elim [-h]`) referring to local hypotheses,
+  -- `gex` are the other exceptions
+  (hs, gex, hex, all_hyps) ← decode_simp_arg_list hs,
+  -- Recall, per the discussion above, we produce `tactic expr` thunks rather than actual `expr`s.
+  let hs := hs.map (λ h, i_to_expr_for_apply h),
+  l ← attr.mmap $ λ a, attribute.get_instances a,
+  let l := l.join,
+  let m := l.map (λ h, mk_const h),
+  -- In order to remove the expressions we need to evaluate the thunks.
+  hs ← (hs ++ m).mfilter $ λ h, (do h ← h, return $ expr.const_name h ∉ gex),
+  let hs := if no_dflt then hs else
+    ([`rfl, `trivial, `congr_fun, `congr_arg].map (λ n, (mk_const n))) ++ hs,
+  hs ← if ¬ no_dflt ∨ all_hyps then do
     ctx ← local_context,
-    return $ hs.append (ctx.filter (λ h, h.local_uniq_name ∉ hex)) -- remove local exceptions
-   else return hs
+    return $ hs.append ((ctx.filter (λ h : expr, h.local_uniq_name ∉ hex)).map return) -- remove local exceptions
+  else return hs,
+  -- Finally, run all of the tactics: any that return an expression without metavariables can safely
+  -- be replaced by a `return` tactic.
+  hs.mmap (λ h, do e ← h, if e.has_meta_var then return h else return (return e))
```

Comment:
My reading of this function is a little superficial, but it looks like you evaluate the thunks twice, once here and once in line 73. This is what you want to avoid, right? You could use `list.filter_map` (or implement `list.mfilter_mmap`? Does that exist?) to do both in one step, so you only evaluate once.

Reply:
I'm actually not worried about evaluating thunks multiple times here: this is where we're building the set of lemmas by parsing user arguments and inspecting the environment. It's only later when we're iteratively calling `apply_any` that we need to worry about doing something over and over again unnecessarily.

Nevertheless, it is unnecessary here. I'll have a look at the plumbing, and either fix it or add a note explaining that it would be an unnecessary optimisation.

Reply:
I've added a note. It seemed a lot of plumbing for little benefit.